packages/astro/src/runtime/server/index.ts

```diff
@@ -462,17 +467,10 @@ export async function renderPage(result: SSRResult, Component: AstroComponentFac
 	if (needsHydrationStyles) {
 		styles.push(renderElement('style', { props: { 'astro-style': true }, children: 'astro-root, astro-fragment { display: contents; }' }));
 	}
-
 	const links = Array.from(result.links)
 		.filter(uniqueElements)
 		.map((link) => renderElement('link', link, false));
-
-	// inject styles & scripts at end of <head>
-	let headPos = template.indexOf('</head>');
-	if (headPos === -1) {
-		return links.join('\n') + styles.join('\n') + scripts.join('\n') + template; // if no </head>, prepend styles & scripts
-	}
-	return template.substring(0, headPos) + links.join('\n') + styles.join('\n') + scripts.join('\n') + template.substring(headPos);
+	return unescapeHTML(links.join('\n') + styles.join('\n') + scripts.join('\n') + '\n' + '<!--astro:head:injected-->');
```

Comment:
We also add a `<!--astro:head:injected-->` comment to flag that we've already injected the links/styles/scripts.

Reply:
Why?

Reply:
I think I understand after reading the code below.